CIP-0069/README.md

```diff
@@ -30,31 +32,60 @@ As it stands the scripts being made on cardano often suffer this problem, and th
+
+I see three major use cases when it comes to running spend scripts without datums:
+
+- Use Case 1: A script that acts as a wallet for users. By having no datum spending the user can send directly from exchanges or have friends send to their smart contract wallet with no datum needed.
+
+- Use Case 2: As a DAO treasury. The funds in this script would be controlled by a DAO and anyone can donate/contribute to the DAO without a datum.
+
+- Use Case 3: Allow dApp protocols to have a claim or withdraw mechanism (similar to Ethereum for tokens sent to a contract without call) for claiming tokens sent without a datum.
+
+I'd be remiss if I didn't mention CIP-0112 which has been expanded to improve native script capabilities to provide an alternative solution for use case 1 and 2. But for use case 3, CIP-0112 does not enable a "claim or withdraw mechanism" for contracts.
+
 ## Specification
 
 ### Removing the datum argument
 
-All the script purposes have a form of `Redeemer -> ScriptContext -> a` except the `Spending` one.
-It has the following form: `Datum -> Redeemer -> ScriptContext -> a`. This is enforced by the Cardano ledger.
+All the script purposes have a form of ```Redeemer -> ScriptContext -> a``` except the Spending one. It has the following form: ```Datum -> Redeemer -> ScriptContext -> a```. This is enforced by the Cardano ledger.
 
-We modify the general signature to be `ScriptArgs -> ScriptContext -> a`.
+We propose to make the following modification:
 
-```hs
-data ScriptArgs =
-    RedeemerOnly Redeemer
-  | RedeemerAndDatum Redeemer Datum
+The signature of all scripts will be ```ScriptContext -> a```.
+The ScriptContext type is now a union type with a variant for each script purpose.
+```
+data ScriptContext
+  = Minting TxInfo Redeemer CurrencySymbol
+  | Spending TxInfo Redeemer (Maybe Datum) TxOutRef
+  | Rewarding TxInfo Redeemer Credential
+  | Certifying TxInfo Redeemer Integer TxCert 
+  | Voting TxInfo Redeemer Voter
+  | Proposing TxInfo Redeemer Integer ProposalProcedure
 ```
+The Datum is optional, which will enable us to allow the execution of spending scripts without a datum. 
+One more change will be needed on the ledger side in order to make the Datum optional for spending scripts.
+The ledger UTXOW rule needs to be relaxed, this ledger rule checks if a utxo has an existing datum if the address's payment credential is a phase 2 validation script.
+
+The ScriptPurpose type used in the Redeemers Map is left the same.
+
+
```

Comment:
I suggest we also specify somewhere here maybe that we need enforce a single argumet to plutus evaluating function, which at the very least means changing type signatures for these two functions for `PlutusV3` in a way that they will accept a single `PLC.Data` agument, instead of `[PLC.Data]`:

* [`evaluateScriptCounting`](https://github.com/IntersectMBO/plutus/blob/7707aab438f7266bd99c0a72e78da7b6f708603c/plutus-ledger-api/src/PlutusLedgerApi/V3.hs#L186-L198)
* [`evaluateScriptRestricting`](https://github.com/IntersectMBO/plutus/blob/7707aab438f7266bd99c0a72e78da7b6f708603c/plutus-ledger-api/src/PlutusLedgerApi/V3.hs#L207-L221)